firstpartydata/first_party_data.go

```diff
@@ -0,0 +1,405 @@
+}
+
+func resolveApp(fpdConfigApp *openrtb2.App, bidRequestApp *openrtb2.App, globalFPD map[string][]byte, openRtbGlobalFPD map[string][]openrtb2.Data, hasGlobalFPD bool, bidderName string) (*openrtb2.App, error) {
+
+	if bidRequestApp == nil && fpdConfigApp == nil {
+		return nil, nil
+	}
+
+	if bidRequestApp == nil && fpdConfigApp != nil {
+		return nil, fmt.Errorf("incorrect First Party Data for bidder %s: App object is not defined in request, but defined in FPD config", bidderName)
+	}
+
+	newApp := *bidRequestApp
+	var err error
+
+	if hasGlobalFPD {
+		//apply global fpd if exists
+		if len(globalFPD[appKey]) > 0 {
+			extData := buildExtData(globalFPD[appKey])
+			if len(newApp.Ext) > 0 {
+				newApp.Ext, err = jsonpatch.MergePatch(newApp.Ext, extData)
+			} else {
+				newApp.Ext = extData
+			}
+		}
+		if len(openRtbGlobalFPD[appContentDataKey]) > 0 {
+			if newApp.Content != nil {
+				contentCopy := *newApp.Content
+				contentCopy.Data = openRtbGlobalFPD[appContentDataKey]
+				newApp.Content = &contentCopy
+			} else {
+				newApp.Content = &openrtb2.Content{Data: openRtbGlobalFPD[appContentDataKey]}
+			}
+		}
+	}
+
+	if fpdConfigApp != nil {
+		//apply bidder specific fpd if present
+		newApp, err = mergeApps(&newApp, fpdConfigApp)
+	}
+
+	return &newApp, err
+}
+
+func mergeApps(originalApp *openrtb2.App, fpdConfigApp *openrtb2.App) (openrtb2.App, error) {
+
+	var err error
+	newApp := *originalApp
+
+	newApp.Name = fpdConfigApp.Name
+	newApp.Bundle = fpdConfigApp.Bundle
+	newApp.Domain = fpdConfigApp.Domain
+	newApp.StoreURL = fpdConfigApp.StoreURL
+	newApp.Cat = fpdConfigApp.Cat
+	newApp.SectionCat = fpdConfigApp.SectionCat
+	newApp.PageCat = fpdConfigApp.PageCat
+	newApp.Ver = fpdConfigApp.Ver
+	newApp.Keywords = fpdConfigApp.Keywords
+
+	if len(fpdConfigApp.Ext) > 0 {
+		if len(originalApp.Ext) > 0 {
+			newApp.Ext, err = jsonpatch.MergePatch(originalApp.Ext, fpdConfigApp.Ext)
+		} else {
+			newApp.Ext = fpdConfigApp.Ext
+		}
+	}
+
+	return newApp, err
+}
+
+func buildExtData(data []byte) []byte {
+	res := make([]byte, 0, len(data))
+	res = append(res, []byte(`{"data":`)...)
+	res = append(res, data...)
+	res = append(res, []byte(`}`)...)
+	return res
+}
+
+//ExtractBidderConfigFPD extracts bidder specific configs from req.ext.prebid.bidderconfig
+func ExtractBidderConfigFPD(reqExt *openrtb_ext.RequestExt) map[openrtb_ext.BidderName]*openrtb_ext.ORTB2 {
+
+	fpd := make(map[openrtb_ext.BidderName]*openrtb_ext.ORTB2)
+	reqExtPrebid := reqExt.GetPrebid()
+	if reqExtPrebid.BidderConfigs != nil {
+		for _, bidderConfig := range *reqExtPrebid.BidderConfigs {
+			for _, bidder := range bidderConfig.Bidders {
+
+				if _, present := fpd[openrtb_ext.BidderName(bidder)]; !present {
+					fpd[openrtb_ext.BidderName(bidder)] = &openrtb_ext.ORTB2{}
+				}
+				//this will overwrite previously set site/app/user.
+				//Last defined bidder-specific config will take precedence
+				fpdBidderData := fpd[openrtb_ext.BidderName(bidder)]
```

Comment:
I'm looking at this logic now and wondering if this is correct. I had reached out to Bret last month about how to handle multiple bidderconfigs for the same bidder and root field (app/site/user) and his guidance was to merge them in order and always prefer source over destination when merging. The conversation resulted in the following being added to the spec:
`All merges should prefer source over destination. e.g. if we're merging a bidder-specific bidderconfig into the ortb2 object and the "site.BLAH" field exists in both, PBS-J would override site.BLAH with the new value.`.

Here we aren't merging but rather just using the last defined bidderconfig for a bidder and root field.
I ran some tests against the Java version and it looks like they are actually just using the first defined bidderconfig instead of merging so I'll reach out to Bret to figure out if this is a bug on their end.